support-figma/extended-layout-plugin/src/design-spec-module.ts

```diff
@@ -0,0 +1,292 @@
+        if (destinationNode) {
+          let list = topLevelReactionComponents.get(topLevelComponentNodeName);
+          if (list) {
+            if (!list.includes(destinationNode)) {
+              list.push(destinationNode);
+              await populateReactionChainAsync(
+                destinationNode,
+                topLevelComponentNodeName,
+                topLevelReactionComponents
+              );
+            }
+          } else {
+            topLevelReactionComponents.set(topLevelComponentNodeName, [
+              destinationNode,
+            ]);
+            await populateReactionChainAsync(
+              destinationNode,
+              topLevelComponentNodeName,
+              topLevelReactionComponents
+            );
+          }
+        }
+      }
+    }
+  }
+
+  var reactions = null;
+  let maybeReactions = node as ReactionMixin;
+  if (maybeReactions.reactions) {
+    reactions = maybeReactions.reactions;
+  }
+  if (reactions != null && reactions.length > 0) {
+    for (const it of reactions) {
+      await handleReactionNodeAsync(it.action);
+      for (const action of it.actions ? it.actions : []) {
+        await handleReactionNodeAsync(action);
+      }
+    }
+  }
+
+  if ((node as any).children) {
+    let parent: ChildrenMixin = node as ChildrenMixin;
+    for (const child of parent.children) {
+      await populateReactionChainAsync(
+        child,
+        topLevelComponentNodeName,
+        topLevelReactionComponents
+      );
+    }
+  }
+}
+
+export function loadClippy(): DesignDocSpec | null {
+  // Get the json plugin data from our root node
+  let clippyData = figma.root.getSharedPluginData(
+    Utils.SHARED_PLUGIN_NAMESPACE,
+    Utils.CLIPPY_JSON_KEY
+  );
+  if (!clippyData) return null;
+
+  /*
+  var reviver = function(key, value) {
+    return value;
+  };
+  */
+
+  try {
+    return JSON.parse(clippyData);
+  } catch (e) {
+    console.log("Error parsing clippy JSON: " + e);
+  }
+  return null;
+}
+
+export async function findTextCustomizationNodesAsync(
+  idToDesignComponentSpecMap: Map<string, DesignComponentSpec>,
+  node: BaseNode,
+  customizedNodeArray: BaseNode[],
+  insideComponentDefn?: DesignComponentSpec
+) {
+  if (!node) return;
+  if (idToDesignComponentSpecMap.has(node.id)) {
+    insideComponentDefn = idToDesignComponentSpecMap.get(node.id);
+  } else if (insideComponentDefn) {
+    // Check to see if our name is one of the customizations, and if so, save the node if it is a text node.
+    for (const customization of insideComponentDefn.customizations) {
+      // We should also look for nodes that start with "#" but don't match a customization
+      // or a top-level component name.
+      if (node.name != customization.node) continue;
+      if (
+        customization.kind == DesignCustomizationKind.Text ||
+        customization.kind == DesignCustomizationKind.TextFunction ||
+        customization.kind == DesignCustomizationKind.ComponentReplacement
```

Comment:
Why ComponentReplacement?

Reply:
When I test with Media, there is a customization for the search field which is a ComponentReplacement. See here: https://github.com/google/automotive-design-compose/blob/031ee2816bc5cece5ab721176b15e9436dcfdcc6/reference-apps/aaos-unbundled/mediacompose/src/main/java/com/android/designcompose/reference/mediacompose/Modules.kt#L217